src/main/java/net/spy/memcached/protocol/ascii/CollectionPipedInsertOperationImpl.java

```diff
@@ -102,6 +102,23 @@ assert getState() == OperationState.READING
       return;
     }
     /* ENABLE_REPLICATION end */
+    /* ENABLE_MIGRATION if */
+    if (hasNotMyKey(line)) {
+      addRedirectSingleKeyOperation(line, key);
+      if (insert.isNotPiped()) {
+        transitionState(OperationState.REDIRECT);
+      } else {
+        // Need to call setNextOpIndex().
+        // Single key operation instance is not be cloned when redirection.
+        // It is reused when being redirection.
+        insert.setNextOpIndex(index);
+        // NOT need to call (index++) because if a line has NOT_MY_KEY
+        // and response string has more lines,
+        // the following all lines should have NOT_MY_KEY too.
+      }
+      return;
+    }
```

Comment:
@uhm0311 
아래 코드와 같이 addRedirectSingleKeyOperation() 호출도 1회만 하면 될 것 같습니다.

```java
    if (hasNotMyKey(line)) {
      if (insert.isNotPiped()) {
        addRedirectSingleKeyOperation(line, key);
        transitionState(OperationState.REDIRECT);
      } else if (!needRedirect()) {
        addRedirectSingleKeyOperation(line, key);
        // Need to call setNextOpIndex().
        // Single key operation instance is not be cloned when redirection.
        // It is reused when being redirection.
        insert.setNextOpIndex(index);
      }
      // NOT need to call (index++) because if a line has NOT_MY_KEY
      // and response string has more lines,
      // the following all lines should have NOT_MY_KEY too.
      return;
    }
```

걱정되는 부분은 NOT_MY_KEY 응답이 온 이후에
NOT_MY_KEY 응답만 계속 와야 하는 데, 그렇지 않으면 어떻게 하죠?

Reply:
> 걱정되는 부분은 NOT_MY_KEY 응답이 온 이후에
NOT_MY_KEY 응답만 계속 와야 하는 데, 그렇지 않으면 어떻게 하죠?

nextOpIndex를 설정하면, 응답에 NOT_MY_KEY 라인이 있고 그 이후에 성공한 응답(ex: STORED)이 한 라인이라도 있을 때 성공 응답에 대해서도 연산을 재시도 하게 되어 있습니다. 이를 피하기 위해서는 nextOpIndex 하나로 재시도를 시작할 collection key를 지정하는 것이 아니라 각 collection key에 대해 재시도를 할지 말지 저장해둬야 합니다.

Reply:
예를 들어 BTreePipedInsert 연산을 수행했고 다음과 같은 response string을 얻었다고 가정합니다.
```
bkey0 : STORED
bkey1 : STORED
bkey2 : NOT_MY_KEY
bkey3 : NOT_MY_KEY
bkey4 : STORED
```

이러면 nextOpIndex는 https://github.com/naver/arcus-java-client/pull/536#issuecomment-1229696326 에서 설명한 것과 같이 bkey2를 가리키게 됩니다. 
그리고 nextOpIndex의 사용처를 보면 다음과 같습니다.

```java
// CollectionPipedInsert.BTreePipedInsert.getAsciiCommand()

      for (i = this.nextOpIndex; i < keySize; i++) {
        Long bkey = keyList.get(i);
        byte[] value = encodedList.get(i);
        setArguments(bb, COMMAND, key, bkey, value.length,
            createOption, (i < keySize - 1) ? PIPE : "");
        bb.put(value);
        bb.put(CRLF);
      }
```

nextOpIndex가 bkey2를 가리키고 있고, for문에서 시작하는 값으로 사용됩니다. single key redirect시 operation instance를 복제하지 않고 재사용하므로 bkey2, bkey3, bkey4가 다시 server cluster에 요청으로 들어갈 것입니다. 그런데 bkey4는 이미 STORED로 성공한 응답이므로 다시 server cluster에 요청하면 오히려 비정상적인 동작이 됩니다.
현재 구현의 nextOpIndex의 사용 로직으로는 이렇게 성공한 bkey4에 대해서도 다시 연산을 시도한다는 문제점이 있습니다. 이러한 현상을 피하기 위해서는 `List<String> retryBKeys`와 같은 자료구조를 이용해 redirect시 다시 시도할 bkey들을 bkey2, bkey3만 저장해둘 필요가 있다는 의미입니다.